finetuner/tuner/base.py

```diff
@@ -210,12 +220,86 @@ def _get_data_loader(
         """Get framework specific data loader from the input data."""
         ...
 
+    @staticmethod
+    def _get_num_batches(docs: 'DocumentSequence', batch_size: int) -> int:
+        """Get the number of batches from a document sequence"""
+        n = len(docs)
+        return n // batch_size + 1 * (n % batch_size)
+
     @abc.abstractmethod
     def _train(self, data: AnyDataLoader):
         """Train the model on given labeled data"""
         ...
 
-    @abc.abstractmethod
-    def _eval(self, data: AnyDataLoader):
-        """Evaluate the model on given labeled data"""
-        ...
+    def _eval(
+        self,
+        query_data: 'DocumentSequence',
+        index_data: Optional['DocumentSequence'],
+        label: str,
+        limit: int,
+        distance: str,
+        batch_size: int,
+        preprocess_fn: Optional['PreprocFnType'] = None,
+        collate_fn: Optional['CollateFnType'] = None,
+    ):
+        """Embed the evaluation data and ompute the evaluation metrics"""
+
+        self.state.num_batches_query = self._get_num_batches(query_data, batch_size)
+        self.state.num_batches_index = (
+            self._get_num_batches(index_data, batch_size) if index_data else 0
+        )
+        self.state.batch_index = 0
+
+        self._trigger_callbacks('on_val_begin')
+        self._trigger_callbacks('on_val_query_begin')
+
+        for idx, batch in enumerate(
+            batch_document_sequence(query_data, size=batch_size)
+        ):
+            self.state.batch_index = idx
+            self._trigger_callbacks('on_val_query_batch_begin')
+            embed(
+                batch,
+                self.embed_model,
+                device=self._device_name,
+                batch_size=batch_size,
+                preprocess_fn=preprocess_fn,
+                collate_fn=collate_fn,
+            )
+            self._trigger_callbacks('on_val_query_batch_end')
+
+        self._trigger_callbacks('on_val_query_end')
+
+        if index_data:
```

Comment:
what is the reason why `index_data` needs to be embedded? 

If I want to evaluate I want to evaluate against something that I am not affecting right?

Reply:
as the model changes during training, index data are represented differently